File: sword_for_offer/codes/java/sfo_36_binary_search_tree_and_doubly_linked_list_s1/sfo_36_binary_search_tree_and_doubly_linked_list_s1.java
/*
* File: sfo_36_binary_search_tree_and_doubly_linked_list_s1.java
* Created Time: 2021-12-09
* Author: Krahets (krahets@163.com)
*/

package sfo_36_binary_search_tree_and_doubly_linked_list_s1;

import include.*;
import java.util.*;

// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;

    public Node(int val) {
        this.val = val;
        this.left = null;
        this.right = null;
    }
};

// ===== Solution Code =====
class Solution {
    Node pre, head;
    public Node treeToDoublyList(Node root) {
        if(root == null) return null;
        dfs(root);
        head.left = pre;
        pre.right = head;
        return head;
    }
    void dfs(Node cur) {
        if(cur == null) return;
        dfs(cur.left);
        if(pre != null) pre.right = cur;
        else head = cur;
        cur.left = pre;
        pre = cur;
        dfs(cur.right);
    }
}

public class sfo_36_binary_search_tree_and_doubly_linked_list_s1 {
    public static void main(String[] args) {
        // ======= Test Case =======
        List<Node> nodeList = new ArrayList<Node>() {{
            add(new Node(1));
            add(new Node(2));
            add(new Node(3));
            add(new Node(4));
            add(new Node(5));
        }};
        nodeList.get(3).left = nodeList.get(1);
        nodeList.get(3).right = nodeList.get(4);
        nodeList.get(1).left = nodeList.get(0);
        nodeList.get(1).right = nodeList.get(2);
        Node root = nodeList.get(3);

        // ====== Driver Code ======
        Solution slt = new Solution();
        Node res = slt.treeToDoublyList(root);
        // Print the Doubly circular linked list
        List<String> nodesVal = new ArrayList<>();
        for (int i = 0; i <= nodeList.size(); i++) {
            nodesVal.add(String.valueOf(res.val));
            res = res.right;
        }
        System.out.println(String.join(" <-> ", nodesVal));
    }
}

File: sword_for_offer/codes/python/sfo_36_binary_search_tree_and_doubly_linked_list_s1.py
'''
File: sfo_36_binary_search_tree_and_doubly_linked_list_s1.py
Created Time: 2021-12-09
Author: Krahets (krahets@163.com)
'''

from include import *

# Definition for a Node.
class Node:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# ===== Solution Code =====
class Solution:
    def treeToDoublyList(self, root: 'Node') -> 'Node':
        def dfs(cur):
            if not cur: return
            dfs(cur.left) # 递归左子树
            if self.pre: # 修改节点引用
                self.pre.right, cur.left = cur, self.pre
            else: # 记录头节点
                self.head = cur
            self.pre = cur # 保存 cur
            dfs(cur.right) # 递归右子树
        
        if not root: return
        self.pre = None
        dfs(root)
        self.head.left, self.pre.right = self.pre, self.head
        return self.head

# ======= Test Case =======
node_list = [Node(i) for i in range(1, 6)]
node_list[3].left = node_list[1]
node_list[3].right = node_list[4]
node_list[1].left = node_list[0]
node_list[1].right = node_list[2]
root = node_list[3]
# ====== Driver Code ======
slt = Solution()
res = slt.treeToDoublyList(root)
# Print the Doubly circular linked list
nodes_val = []
for _ in range(len(node_list) + 1):
    nodes_val.append(str(res.val))
    res = res.right
print(' <-> '.join(nodes_val))
